apm-lambda-extension/apmproxy/receiver.go

```diff
@@ -121,17 +122,36 @@ func (c *Client) handleIntakeV2Events() func(w http.ResponseWriter, r *http.Requ
 			return
 		}
 
-		if len(rawBytes) > 0 {
-			agentData := AgentData{
-				Data:            rawBytes,
-				ContentEncoding: r.Header.Get("Content-Encoding"),
-			}
+		flushed := r.URL.Query().Get("flushed") == "true"
 
-			c.EnqueueAPMData(agentData)
+		agentData := AgentData{
+			Data:            rawBytes,
+			ContentEncoding: r.Header.Get("Content-Encoding"),
+			Flushed:         flushed,
 		}
 
-		if len(r.URL.Query()["flushed"]) > 0 && r.URL.Query()["flushed"][0] == "true" {
-			c.AgentDoneSignal <- struct{}{}
+		enqueued := c.EnqueueAPMData(agentData)
+
+		if enqueued && flushed {
```

Comment:
I missed this before: this means that we will not end `processEvent` early if the agent has indicated it has flushed, but the extension's queue is full. That doesn't seem ideal.

Perhaps instead of introducing `AgentData.Flushed` and recreating the channel when processing data, we could add a method `Client.ResetFlushed()`:

```go
// ResetFlushed resets the client's "agent flushed" state, such that
// subsequent calls to WaitForFlush will block until another request
// is received from the agent indicating it has flushed.
func (c *Client) ResetFlushed() {
    c.flushMutex.Lock()
    defer c.flushMutex.Unlock()
    c.flushCh = make(chan struct{})
}
```

Then we would call that at the end of `App.processEvent`? WDYT?

It's still not perfect, becaause `App.processEvent` could end due to the not-quite-timeout being reached. It's still possible that that timer fires but a `?flushed=true` request is then received, closing the channel after it's reset. I think we'll need to make more radical changes to fix that - later.

Reply:
Agree, we can focus on the linked bug here and tackle the other issues in a followup discussion//PR.

I liked `ResetFlushed`, I've updated the code